rf_tests.go

```diff
@@ -362,6 +369,108 @@ func uploadRFML(c cliContext, api rfmlAPI) error {
 	if err != nil {
 		return cli.NewExitError(err.Error(), 1)
 	}
+
+	wisp_tests, err := readWispFiles([]string{c.String("test-folder")})
+	if err != nil {
+		return cli.NewExitError(err.Error(), 1)
+	}
+	err = uploadWispFiles(wisp_tests, false, api)
+	if err != nil {
+		return cli.NewExitError(err.Error(), 1)
+	}
+
+	return nil
+}
+
+// uploadSingleWispFile uploads wisp file syntax by
+// trying to parse the file and sending any parse errors to the caller
+func uploadSingleWispFile(filePath string) error {
+	wisp, _ := readWispFile(filePath)
+
+	log.Printf("Updating existing wisp test: %v", wisp.TestID)
+	err := api.UpdateWisp(wisp)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func readWispFile(filePath string) (*rainforest.WispJson, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	wispReader := rainforest.NewWispReader(file)
+	wisp, err := wispReader.Read()
+	if err != nil {
+		return nil, fileParseError{filePath, err}
+	}
+
+	return wisp, err
+}
+
+// readWispFiles takes in a list of files and/or directories and a list of tags
+// and returns a list of the parsed tests, or an error if it is encountered. To
+// allow all tags, pass in nil for tags.
+func readWispFiles(files []string) ([]*rainforest.WispJson, error) {
+	fileList := []string{}
+	for _, file := range files {
+		stat, err := os.Stat(file)
+		if err != nil {
+			return nil, err
+		}
+		if !stat.IsDir() {
+			if strings.HasSuffix(file, ".json") {
+				fileList = append(fileList, file)
+				continue
+			} else {
+				log.Printf("%s is not a valid json file", file)
+				continue
+			}
+		}
+
+		// We have a directory, walk through and find wisp files
+		err = filepath.Walk(file, func(path string, f os.FileInfo, err error) error {
+			if strings.HasSuffix(path, ".json") {
+				fileList = append(fileList, path)
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	tests := []*rainforest.WispJson{}
+	seenPaths := map[string]bool{}
+	for _, filePath := range fileList {
+		// No dups!
+		if seenPaths[filePath] {
+			continue
+		}
+		seenPaths[filePath] = true
+		wisp, err := readWispFile(filePath)
+		if err != nil {
+			return nil, err
+		}
+		tests = append(tests, wisp)
+	}
+	return tests, nil
+}
```

Comment:
I think this function is doing too much as it's finding all the Wisp looking tests and it's also parsing them?

Perhaps split it into two function, the first does the file discovery and returns a list of (unique) paths, the second maps over this list of paths and returns an array of `WispJson` objects.

Reply:
I ended up separating intro three functions. One finds the files, second removes dups and third one parses the json.
I plan to further refactor this in a later PR to reuse the dup filter and probably the file discovery as well for rfml and wisp to reduce code duplication.